modules/light-clients/08-wasm/types/migrate_contract_test.go

```diff
@@ -0,0 +1,140 @@
+				err := ibcwasm.CodeHashes.Set(suite.chainA.GetContext(), newHash[:])
+				suite.Require().NoError(err)
+
+				payload = []byte{1}
+				expChecksum := wasmvmtypes.ForceNewChecksum(hex.EncodeToString(newHash[:]))
+
+				suite.mockVM.MigrateFn = func(codeID wasmvm.Checksum, env wasmvmtypes.Env, msg []byte, _ wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					suite.Require().Equal(expChecksum, codeID)
+					suite.Require().Equal(defaultWasmClientID, env.Contract.Address)
+					suite.Require().Equal(payload, msg)
+
+					data, err := json.Marshal(types.EmptyResult{})
+					suite.Require().NoError(err)
+
+					return &wasmvmtypes.Response{Data: data}, wasmtesting.DefaultGasUsed, nil
+				}
+			},
+			nil,
+		},
+		{
+			"success: update client state",
+			func() {
+				newHash = sha256.Sum256([]byte{1, 2, 3})
+				err := ibcwasm.CodeHashes.Set(suite.chainA.GetContext(), newHash[:])
+				suite.Require().NoError(err)
+
+				suite.mockVM.MigrateFn = func(_ wasmvm.Checksum, _ wasmvmtypes.Env, _ []byte, store wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					expClientState = types.NewClientState([]byte{1}, newHash[:], clienttypes.NewHeight(2000, 2))
+					store.Set(host.ClientStateKey(), clienttypes.MustMarshalClientState(suite.chainA.App.AppCodec(), expClientState))
+
+					data, err := json.Marshal(types.EmptyResult{})
+					suite.Require().NoError(err)
+
+					return &wasmvmtypes.Response{Data: data}, wasmtesting.DefaultGasUsed, nil
+				}
+			},
+			nil,
+		},
+		{
+			"failure: new and old code hash are the same",
+			func() {
+				// this should not be called
+				suite.mockVM.MigrateFn = func(_ wasmvm.Checksum, _ wasmvmtypes.Env, _ []byte, _ wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					panic("unreachable")
+				}
+			},
+			types.ErrWasmCodeExists,
+		},
+		{
+			"failure: code hash not found",
+			func() {
+				newHash = sha256.Sum256([]byte{1, 2, 3})
+			},
+			types.ErrWasmCodeHashNotFound,
+		},
+		{
+			"failure: contract returns error",
+			func() {
+				newHash = sha256.Sum256([]byte{1, 2, 3})
+				err := ibcwasm.CodeHashes.Set(suite.chainA.GetContext(), newHash[:])
+				suite.Require().NoError(err)
+
+				suite.mockVM.MigrateFn = func(_ wasmvm.Checksum, _ wasmvmtypes.Env, _ []byte, _ wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.Response, uint64, error) {
+					return nil, wasmtesting.DefaultGasUsed, wasmtesting.ErrMockContract
+				}
+			},
+			types.ErrWasmContractCallFailed,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupWasmWithMockVM()
+
+			oldHash = sha256.Sum256(wasmtesting.Code)
+			newHash = oldHash
```

Comment:
I think it makes more sense to actually have the setting of the new code hash before the malleate and have this in the failure case. wdyt?

Reply:
+1

Reply:
done